### PR TITLE
Raise a neater RuntimeError when the correct async deps are not installed.

### LINKED ISSUE #825: AttributeError: 'AsyncLock' object has no attribute '_anyio_lock'
Try install httpx + httpcore 1.0 then remove anyio. Error:
```
(.venv) E:\hcore1>python app.py
Traceback (most recent call last):
  File "E:\hcore1\app.py", line 6, in main
    r = await client.get('https://www.example.com/')
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_client.py", line 1757, in get
    return await self.request(
           ^^^^^^^^^^^^^^^^^^^
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_client.py", line 1530, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_client.py", line 1617, in send
    response = await self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_client.py", line 1645, in _send_handling_auth
    response = await self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_client.py", line 1682, in _send_handling_redirects
    response = await self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_client.py", line 1719, in _send_single_request
    response = await transport.handle_async_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_transports\default.py", line 366, in handle_async_request
    resp = await self._pool.handle_async_request(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\hcore1\.venv\Lib\site-packages\httpcore\_async\connection_pool.py", line 224, in handle_async_request
    async with self._pool_lock:
  File "E:\hcore1\.venv\Lib\site-packages\httpcore\_synchronization.py", line 62, in __aenter__
    self.setup()
  File "E:\hcore1\.venv\Lib\site-packages\httpcore\_synchronization.py", line 55, in setup
    raise RuntimeError(
RuntimeError: Running with asyncio requires installation of 'httpcore[asyncio]'.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\hcore1\app.py", line 8, in <module>
    asyncio.run(main())
  File "C:\Users\User\.rye\py\cpython@3.12.0\install\Lib\asyncio\runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\User\.rye\py\cpython@3.12.0\install\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\.rye\py\cpython@3.12.0\install\Lib\asyncio\base_events.py", line 664, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "E:\hcore1\app.py", line 5, in main
    async with httpx.AsyncClient() as client:
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_client.py", line 2003, in __aexit__
    await self._transport.__aexit__(exc_type, exc_value, traceback)
  File "E:\hcore1\.venv\Lib\site-packages\httpx\_transports\default.py", line 345, in __aexit__
    await self._pool.__aexit__(exc_type, exc_value, traceback)
  File "E:\hcore1\.venv\Lib\site-packages\httpcore\_async\connection_pool.py", line 327, in __aexit__
    await self.aclose()
  File "E:\hcore1\.venv\Lib\site-packages\httpcore\_async\connection_pool.py", line 312, in aclose
    async with self._pool_lock:
  File "E:\hcore1\.venv\Lib\site-packages\httpcore\_synchronization.py", line 67, in __aenter__
    await self._anyio_lock.acquire()
          ^^^^^^^^^^^^^^^^
AttributeError: 'AsyncLock' object has no attribute '_anyio_lock'

(.venv) E:\hcore1>
```

_Originally posted by @T-256 in https://github.com/encode/httpx/pull/2885#discussion_r1356548422_

### PR DESCRIPTION
**Pull request by tomchristie**
Closes #825.

**example.py**

```python
import httpcore
import asyncio


async def main():
    async with httpcore.AsyncConnectionPool() as pool:
        response = await pool.request("GET", "https://www.example.com")
    print(response)


asyncio.run(main())
```

**output:**

```shell
$ venv/bin/python ./example.py 
Traceback (most recent call last):
  File "/Users/tomchristie/Temp/./example.py", line 11, in <module>
    asyncio.run(main())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/tomchristie/Temp/./example.py", line 6, in main
    async with httpcore.AsyncConnectionPool() as pool:
  File "/Users/tomchristie/GitHub/encode/httpcore/httpcore/_async/connection_pool.py", line 331, in __aenter__
    async with self._pool_lock:
  File "/Users/tomchristie/GitHub/encode/httpcore/httpcore/_synchronization.py", line 64, in __aenter__
    self.setup()
  File "/Users/tomchristie/GitHub/encode/httpcore/httpcore/_synchronization.py", line 56, in setup
    self._backend = current_async_library()
  File "/Users/tomchristie/GitHub/encode/httpcore/httpcore/_synchronization.py", line 35, in current_async_library
    raise RuntimeError(
RuntimeError: Running with asyncio requires installation of 'httpcore[asyncio]'.
```

Compare this against... https://github.com/encode/httpcore/issues/825#issuecomment-1761296515

### CODE REVIEW

File: httpcore/_synchronization.py [hunk 1]
```diff
@@ -201,24 +175,14 @@ def __init__(self) -> None:
...
-
+        elif self._backend == "asyncio":
             self._anyio_shield = anyio.CancelScope(shield=True)
 
     def __enter__(self) -> "AsyncShieldCancellation":
         if self._backend == "trio":
             self._trio_shield.__enter__()
-        else:
+        elif self._backend == "asyncio":
             self._anyio_shield.__enter__()
```
Comment (zanieb) [inline httpcore/_synchronization.py:186]: Should we silently return `self` without doing anything if the backend is somehow invalid? Should there be an `else` that errors instead?

Comment (T-256) [reply httpcore/_synchronization.py:186]: Or using `AsyncBackend` enum is proper? we sure only trio and asyncio are acceptable backend.
```py
self._backend: Optional[AsyncBackend] = None
```

File: httpcore/_async/connection_pool.py [hunk 2]
```diff
@@ -326,6 +326,10 @@ async def aclose(self) -> None:
             self._requests = []
 
     async def __aenter__(self) -> "AsyncConnectionPool":
+        # Acquiring the pool lock here ensures that we have the
+        # correct dependencies installed as early as possible.
+        async with self._pool_lock:
+            pass
```
Comment (T-256) [inline httpcore/_async/connection_pool.py:332]: I think here don't need extra await with immediately lock release.

```suggestion
        # Ensure correct dependencies installed as early as possible.
        if not self._pool_lock._backend:
            self._pool_lock.setup()
```

**Review by T-256 — APPROVED** [review]

File: httpcore/_async/connection_pool.py (hunk 2 above)
Comment (tomchristie) [reply httpcore/_async/connection_pool.py:332]: I'd generally prefer using the lock's public API here.

File: httpcore/_synchronization.py (hunk 1 above)
Comment (tomchristie) [reply httpcore/_synchronization.py:186]: I think it's *probably* okay enough as-is?
Could we pull this in now and deal with any alternate suggestions as a follow up?

Comment (T-256) [reply httpcore/_synchronization.py:186]: Agree, works well and it doesn't disrupt anything btw.

**Comment by chamikabm** [comment]
Any plan on releasing this fix ?

**Review by zanieb — APPROVED** [review]

**Pull request merged**